### ISSUE TITLE
Class seen as not hashable even if the parent defines the relevant methods

### ISSUE THREAD

**Issue by guillaume-alliander** [microsoft/pyright#7344]
**Describe the bug**
Class seen as not hashable even if the parent defines the relevant methods.

Weirdly enough, if my parent class in the example does not inherit from `BaseModel` all is fine. But as using `Parent` directly makes pyright happy, I still think that something happens in pyright. 

**Code or Screenshots**

```python
from pydantic import BaseModel # pydantic==2.6.1 pydantic_core==2.16.2


class Parent(BaseModel):
    ID: int = 1

    def __hash__(self) -> int:
        return self.ID

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, type(self)):
            return NotImplemented
        return self.__hash__() == other.__hash__()


class Child(Parent):
    pass


works = {Parent()}  # hashable, pyright happy
fails = {Child()}  # allegedly not hashable, pyright unhappy
```

**VS Code extension or command-line**
Command line:
pyright --version
pyright 1.1.351

But I see the same behaviour in vscode.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. 

As a standards-based type checker, pyright has no specialized knowledge of the behaviors of pydantic or other third-party libraries. It works entirely from the type information it is provided by those libraries. Pydantic is leveraging the `dataclass_transform` mechanism defined in [PEP 681](https://peps.python.org/pep-0681/). It tells type checkers to treat classes that derive from pydantic's `BaseMethod` to act like the stdlib dataclass.

The stdlib dataclass does not treat subclasses as hashable unless they are frozen. This makes sense because subclasses typically add new fields, and if these fields are mutable (not frozen), then the hashes of those fields would change.

Because pydantic's behavior here differs from stdlib dataclass, you'll need to find a workaround if you want your code to work with pyright or other type checkers that support `dataclass_transform`.

**Comment by guillaume-alliander** [comment]
Thanks for the extended explanation (and for a great tool)!
